pkg/controller/job/job_controller.go

```diff
@@ -1917,10 +1917,10 @@ func (jm *Controller) cleanupPodFinalizers(job *batch.Job) {
 	}
 }
 
-func recordJobPodsCreationTotal(job *batch.Job, succeeded, failed int32) {
+func recordJobPodsCreationTotal(job *batch.Job, terminating *int32, succeeded, failed int32) {
 	reason := metrics.PodCreateNew
 	if feature.DefaultFeatureGate.Enabled(features.JobPodReplacementPolicy) {
-		podsTerminating := job.Status.Terminating != nil && *job.Status.Terminating > 0
+		podsTerminating := ptr.Deref(job.Status.Terminating, 0) > 0 || ptr.Deref(terminating, 0) > 0
```

Comment:
Actually, why should we check `job.Status`? Isn't that information stale at this point? should we just compare the values of `jobCtx` only?

Reply:
Good point, we don't need to use job.Status.Terminating. However, when thought about this, there could be a period when the pods transition from terminating -> uncounted -> failed. When they are uncounted the current implementation would not see them and assume "new". To fix that I propose to use the jobCtx.failed as well, which includes the uncounted pods.

Reply:
I had to adjust unit tests by defaulting podReplacementPolicy, because the tests would panic, as it would often enter the check `*job.Spec.PodReplacementPolicy == batch.Failed`, because `jobCtx.failed=true` or `jobCtx.terminating>0` before this is reflected in status.

I also made a change that is not strictly required for improving performance, but fits into "improve accuracy of the metric", that when `*job.Spec.PodReplacementPolicy == batch.Failed` then I only check if `jobCtx.failed>0`, this is because terminating pods would not trigger replacement. So, when `terminating>0 and failed=0` I think we still have "new" pod creations.

PTAL